minetest-gymnasium/minetest/minetest_env.py

```diff
@@ -108,32 +110,97 @@ def _free_port():
         return s.getsockname()[1]
 
 
+class Observation(TypedDict):
+    image: np.typing.NDArray[np.uint8]
+
+
+class HudElement(TypedDict):
+    name: str
+    text: str
+
+
+class Info(TypedDict):
+    hud_elements: list[HudElement]
+    player_health: int
+    player_health_max: int
+    player_breath: int
+    player_breath_max: int
+    player_is_dead: bool
+    player_metadata: dict[str, str]
+
+
+RewardFn = Callable[[Observation, Info], float]
```

Comment:
An alternative to this is to always return zero and implement wrappers for each game that set the reward. That seems a bit nicer to me in terms of layering things.

Same goes for the observation spaces. The base class here could just return an image and everything else in the info dict, and the wrapper could parse things from info into the obs.

WDYT?

https://gymnasium.farama.org/tutorials/gymnasium_basics/implementing_custom_wrappers/#inheriting-from-gymnasium-wrapper

Reply:
I deleted the additional observation spaces exactly because it can be done because of a wrapper. I think it would be better too but re-added this because the different wrappers would have to do similar things.

I suggest we keep this for now to facilitate upgrading the code. We refactor and provide minetest-boad and minetest-globo environments (in the mod repos?) which can use obs and action wrappers. Then we can delete the additional observation spaces.

Reply:
#87